pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1559,33 +1560,32 @@ public void addTable(TableConfig tableConfig)
     if (ZKMetadataProvider.getSchema(_propertyStore, TableNameBuilder.extractRawTableName(tableNameWithType)) == null) {
       throw new InvalidTableConfigException("No schema defined for table: " + tableNameWithType);
     }
-    if (tableType == TableType.OFFLINE) {
-      try {
-        // Add table config
-        ZKMetadataProvider.setTableConfig(_propertyStore, tableNameWithType, TableConfigUtils.toZNRecord(tableConfig));
-        // Assign instances
-        assignInstances(tableConfig, true);
+    Preconditions.checkState(tableType == TableType.OFFLINE || tableType == TableType.REALTIME,
+        "Invalid table type: %s", tableType);
+    try {
+      // Add table config
+      if (!ZKMetadataProvider.createTableConfig(_propertyStore, tableNameWithType, tableConfig)) {
+        if (getTableConfig(tableNameWithType) != null) {
```

Comment:
We already have this check on line 1542, so IMO we can directly throw the `RuntimeException`